### PR TITLE
SQLInstance: handle retired maintenance versions in field defaults

### PR DESCRIPTION
**Pull request by jingyih**
Only default `spec.maintenanceVersion` to `maintenanceVersion` from live state if the version is still valid.

When a user doesn't specify a `maintenanceVersion` in their SQLInstance spec, the KCC controller defaults to using the version from the live state. However, the Cloud SQL service may retire older maintenance versions. When the version on the live instance was retired, KCC would continue to send it during subsequent updates, causing the API to reject the requests and putting the resource stuck in UpdateFailed state.

```
message: 'Update call failed: error updating: updating SQLInstance xxx-yyy-zzz  
      failed: googleapi: Error 400: Invalid request: The requested maintenance version  
      is not an available maintenance version to upgrade. Please refer to the availableMaintenanceVersions  
      on the instance resource to see the available maintenance versions.., invalid' 
```

### CODE REVIEW

**Comment by jingyih** [comment]
/hold
After 1.132.1 is released.

File: pkg/controller/direct/sql/sqlinstance_defaults.go [hunk 1]
```diff
@@ -22,14 +22,58 @@ import (
 	api "google.golang.org/api/sqladmin/v1beta4"
 )
 
+// isMaintenanceVersionAvailable checks if the given maintenance version is available in the list of available versions.
+func isMaintenanceVersionAvailable(version string, availableVersions []string) bool {
+	if version == "" || len(availableVersions) == 0 {
```
Comment (maqiuyujoyce) [inline pkg/controller/direct/sql/sqlinstance_defaults.go:27]: In line 55, `len(availableVersions) == 0` means the maintenanceVersion _is_ available. And in line 47, `actualVersion == ""` means the we'll return as is. But the behavior of these two conditions are different here. I feel this might confuse future contributors who maintain the code. Could you add some comments to clarify the behavioral differences?

File: pkg/controller/direct/sql/sqlinstance_defaults.go [hunk 2]
```diff
@@ -22,14 +22,58 @@ import (
...
+			return true
+		}
+	}
+	return false
+}
+
+// determineMaintenanceVersion determines the appropriate maintenance version to use based on user specification and actual instance state.
+// Returns the maintenance version to use, or empty string if none should be set.
+func determineMaintenanceVersion(userSpecifiedVersion *string, actualVersion string, availableVersions []string) string {
+	// If user specified a version, use it regardless of actual/available
```
Comment (maqiuyujoyce) [inline pkg/controller/direct/sql/sqlinstance_defaults.go:41]: Is it possible that: *userSpecifiedVersion is expired, but the API updates the actual maintenance version during creation (so *userSpecifiedVersion != actualVersion); then during the update, the API errors out when *userSpecifiedVersion != actualVersion?

I'm brainstorming the edge cases because the API behaviors can be really unpredictable.

File: pkg/controller/direct/sql/sqlinstance_defaults.go [hunk 3]
```diff
@@ -22,14 +22,58 @@ import (
...
+
 func ApplySQLInstanceGCPDefaults(in *krm.SQLInstance, out *api.DatabaseInstance, actual *api.DatabaseInstance) {
 	if in.Spec.InstanceType == nil {
 		// GCP default InstanceType is CLOUD_SQL_INSTANCE.
 		out.InstanceType = "CLOUD_SQL_INSTANCE"
 	}
 	if in.Spec.MaintenanceVersion == nil && actual != nil {
-		// If desired maintenanceVersion is not specified, assume user wants the actual.
-		out.MaintenanceVersion = actual.MaintenanceVersion
+		maintenanceVersion := determineMaintenanceVersion(in.Spec.MaintenanceVersion, actual.MaintenanceVersion, actual.AvailableMaintenanceVersions)
```
Comment (maqiuyujoyce) [inline pkg/controller/direct/sql/sqlinstance_defaults.go:73]: From the code logic in determineMaintenanceVersion(), it should also apply to `in.Spec.MaintenanceVersion` when it's not nil, right?

File: pkg/controller/direct/sql/sqlinstance_defaults_test.go [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+package sql
+
+import (
+	"testing"
+
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/direct"
+)
+
+func TestDetermineMaintenanceVersion(t *testing.T) {
```
Comment (maqiuyujoyce) [inline pkg/controller/direct/sql/sqlinstance_defaults_test.go:23]: Thank you for adding the test cases!

**Review by maqiuyujoyce — COMMENTED** [review]
Can we have a scenario test for the expired version edge case as well?

File: pkg/controller/direct/sql/sqlinstance_defaults.go (hunk 2 above)
Comment (jingyih) [reply pkg/controller/direct/sql/sqlinstance_defaults.go:41]: This PR only handles "defaulting". If users set `userSpecifiedVersion` but it became invalid later, they need to update the value of `userSpecifiedVersion` in their YAML.

File: pkg/controller/direct/sql/sqlinstance_defaults.go (hunk 3 above)
Comment (jingyih) [reply pkg/controller/direct/sql/sqlinstance_defaults.go:73]: If I understand your question correctly, yes. We will use the value from `in.Spec.MaintenanceVersion` if it is set by user. So in this case, there is no defaulting.

File: pkg/controller/direct/sql/sqlinstance_defaults.go (hunk 1 above)
Comment (jingyih) [reply pkg/controller/direct/sql/sqlinstance_defaults.go:27]: I don't quite follow your comment. But I do have a unit test for function "determineMaintenanceVersion" to describe its behavior.

**Comment by maqiuyujoyce** [comment]
> Only default `spec.maintenanceVersion` to `maintenanceVersion` from live state if the version is still valid.
> 
> When a user doesn't specify a `maintenanceVersion` in their SQLInstance spec, the KCC controller defaults to using the version from the live state. However, the Cloud SQL service may retire older maintenance versions. When the version on the live instance was retired, KCC would continue to send it during subsequent updates, causing the API to reject the requests and putting the resource stuck in UpdateFailed state.
> 
> ```
> message: 'Update call failed: error updating: updating SQLInstance xxx-yyy-zzz  
>       failed: googleapi: Error 400: Invalid request: The requested maintenance version  
>       is not an available maintenance version to upgrade. Please refer to the availableMaintenanceVersions  
>       on the instance resource to see the available maintenance versions.., invalid' 
> ```

@jingyih If the original issue isn't the root cause, then do you happen to know what caused this error?

**Comment by maqiuyujoyce** [comment]
Thank you, @jingyih ! That makes sense.
